File: com/dev/systemdesign/InMemoryKVStoreWithTTL.java
/*
In-memory key-value store where entries expire after a TTL, supports lazy expiry on
read plus an active purge sweep, common system design interview question.

status - completed
 */

package com.dev.systemdesign;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import com.dev.logger.basePrinter;

class InMemoryKVStoreWithTTL extends basePrinter{

    private static class Entry {
        String value;
        long expiresAtMillis; // -1 means never expires

        Entry(String value, long expiresAtMillis){
            this.value = value;
            this.expiresAtMillis = expiresAtMillis;
        }

        boolean isExpired(long now){
            return expiresAtMillis != -1 && now >= expiresAtMillis;
        }
    }

    private final HashMap<String, Entry> store;

    InMemoryKVStoreWithTTL(){
        this.store = new HashMap<>();
    }

    void put(String key, String value){
        store.put(key, new Entry(value, -1));
    }

    void put(String key, String value, long ttlMillis){
        store.put(key, new Entry(value, System.currentTimeMillis() + ttlMillis));
    }

    String get(String key){
        Entry entry = store.get(key);
        if (entry == null){
            return null;
        }
        if (entry.isExpired(System.currentTimeMillis())){
            store.remove(key); // lazy expiry
            return null;
        }
        return entry.value;
    }

    int purgeExpired(){
        int removed = 0;
        long now = System.currentTimeMillis();
        Iterator<Map.Entry<String, Entry>> it = store.entrySet().iterator();
        while (it.hasNext()){
            Map.Entry<String, Entry> mapEntry = it.next();
            if (mapEntry.getValue().isExpired(now)){
                it.remove();
                removed++;
            }
        }
        return removed;
    }

    int size(){
        return store.size();
    }

    private static void sleep(long millis){
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public static void main(String[] args) {
        InMemoryKVStoreWithTTL kvStore = new InMemoryKVStoreWithTTL();
        kvStore.put("permanent", "stays forever");
        kvStore.put("shortLived", "expires soon", 200);

        logp("get(permanent) -> " + kvStore.get("permanent"));
        logp("get(shortLived) immediately -> " + kvStore.get("shortLived"));

        sleep(300);

        logp("get(shortLived) after ttl elapsed (lazy expiry) -> " + kvStore.get("shortLived"));
        logp("get(permanent) still present -> " + kvStore.get("permanent"));

        kvStore.put("anotherShortLived", "also expires", 100);
        logp("size before purge -> " + kvStore.size());
        sleep(200);
        int purged = kvStore.purgeExpired();
        logp("entries purged -> " + purged);
        logp("size after purge -> " + kvStore.size());
    }
}

File: typescript/systemdesign/inMemoryKVStoreWithTTL.ts
/*
In-memory key-value store where entries expire after a TTL. Expiry is lazy
(checked on read) plus an active purge sweep is available to reclaim memory
from keys nobody has touched since expiring.

status - completed
*/

import { logp } from "../utils/logger";

interface Entry<V> {
  value: V;
  expiresAtMs: number | null;
}

class InMemoryKVStoreWithTTL<V> {
  private readonly store: Map<string, Entry<V>> = new Map();

  constructor(private readonly nowFn: () => number = Date.now) {}

  private isExpired(entry: Entry<V>, nowMs: number): boolean {
    return entry.expiresAtMs !== null && entry.expiresAtMs <= nowMs;
  }

  set(key: string, value: V, ttlMs?: number): void {
    const nowMs = this.nowFn();
    const expiresAtMs = ttlMs !== undefined ? nowMs + ttlMs : null;
    this.store.set(key, { value, expiresAtMs });
  }

  get(key: string): V | undefined {
    const entry = this.store.get(key);
    if (!entry) {
      return undefined;
    }
    if (this.isExpired(entry, this.nowFn())) {
      this.store.delete(key);
      return undefined;
    }
    return entry.value;
  }

  has(key: string): boolean {
    return this.get(key) !== undefined;
  }

  delete(key: string): boolean {
    return this.store.delete(key);
  }

  // active purge: sweeps all entries and evicts anything already expired,
  // without waiting for a read to trigger lazy expiry
  purgeExpired(): number {
    const nowMs = this.nowFn();
    let purged = 0;
    for (const [key, entry] of this.store) {
      if (this.isExpired(entry, nowMs)) {
        this.store.delete(key);
        purged++;
      }
    }
    return purged;
  }

  size(): number {
    return this.store.size;
  }
}

// usage scenarios: fake clock for deterministic timing

let simulatedNow = 0;
const kv = new InMemoryKVStoreWithTTL<string>(() => simulatedNow);

kv.set("permanent", "stays-forever");
kv.set("short-lived", "gone-soon", 100);

logp(`get(permanent) -> ${kv.get("permanent")}`);
logp(`get(short-lived) at t=0 -> ${kv.get("short-lived")}`);

simulatedNow = 150;
logp(`get(short-lived) at t=150 (lazy expiry) -> ${kv.get("short-lived")} (expected undefined, expired)`);
logp(`store size after lazy expiry removed it -> ${kv.size()}`);

kv.set("also-expires", "x", 50);
kv.set("keeps-going", "y", 10_000);
simulatedNow = 300;
const purgedCount = kv.purgeExpired();
logp(`active purge at t=300 removed ${purgedCount} expired entr${purgedCount === 1 ? "y" : "ies"}`);
logp(`get(keeps-going) -> ${kv.get("keeps-going")}`);
